### PR TITLE
[elastic] Add support for the full edefinition API.

### PR DESCRIPTION
**Pull request by movie-travel-code**
This change mainly has two parts:

 - For the cross definition, collect the destination repo revision.
 - For the cross definition, collect the destination repo URI.

For the cross repo jump, the lsp server will return:
```
SymbolLocator: {
  Qname: ...
  SymbolKind: ...
  Path: relative file path under the repo
  Loc: 'repoURI + 'blob'/revision+ Path'
  PackageLocator: {
    Version: semver format tag like 'v1.2.0'
                        or
             prefix of SHA, like 'abcdef1234567'
    Name: ...
    RepoURI: repo URL
  }
}
```

A typical result:
```
result":[{
  "Qname":"jsoniter.Marshal",
  "Kind":12,
  "Path":"adapter.go",
  "location":{
    "uri":"github.com/json-iterator/go/blob/v1.1.6/adapter.go",
    "range":{"start":{"line":32,"character":5},
    "end":{"line":32,"character":12}}
  },
  "Package":{
    "Version":"v1.1.6",
    "Name":"jsoniter",
    "uri":"github.com/json-iterator/go"
  }
}]
```

Potential issue:
There is no API to get the repo URI for golang, to more precise, golang
is unaware of the repo. So we have to access GitHub to get the repo URI.
Accessing github.com is time-consuming, use cache in the follow-up
change.

### CODE REVIEW

**Comment by movie-travel-code** [comment]
@zfy0701 @spacedragon For now, cross repo jump is still problematic. Commit an initial implementation to help guys to find out where the problem is. The cross repo jump URL and the corresponding correct URL as the figure below:
Screen Shot 2019-04-29 at 11 15 08 PM

@Poytr1 Please review.

**Comment by zfy0701** [comment]
the left seems right? you need add the revision of your repo,  you could extra version from the workspace path

File: internal/lsp/elasticserver.go [hunk 1]
```diff
@@ -269,3 +294,197 @@ func getQName(ctx context.Context, f source.File, ident *source.IdentifierInfo,
...
+	}
+}
+
+func checkGithubRepo(cfg *packages.Config, repoURI string) (string, string) {
+	client := github.NewClient(nil)
+	strs := strings.Split(repoURI, string(filepath.Separator))
+	for i := len(strs) - 1; i >= 0; i-- {
+		if i > 1 {
+			if _, response, err := client.Repositories.Get(cfg.Context, strs[i-1], strs[i]); err == nil {
+				if response.StatusCode == 200 {
```
Comment (Poytr1) [inline internal/lsp/elasticserver.go]: If repo is private, here the status code will be 404

Comment (zfy0701) [reply internal/lsp/elasticserver.go]: @movie-travel-code I don't think we should use github client, let's begin with something simple, just parse the HTTP

calling to remote will raise multiple issues like the one @Poytr1 just mentioned

Comment (movie-travel-code) [reply internal/lsp/elasticserver.go]: @Poytr1 This approach is inextensible, I dig into the go library and find another solution. @zfy0701 You are right, we can parse the import path. Go have a series of regexp patterns to parse the import path for different hosting sites and different VCS scheme.

**Comment by movie-travel-code** [comment]
@zfy0701 The new patch can enable the cross repo jump, see https://github.com/elastic/code/issues/1124#issuecomment-488228161.

**Comment by movie-travel-code** [comment]
> Can this support private repo?

For now, the current approach uses regexp patterns to parse the import path to get the repo URL. And we won't be bothered by whether the repo is private.

File: internal/lsp/elasticserver.go (hunk 1 above)
Comment (movie-travel-code) [reply internal/lsp/elasticserver.go]: @zfy0701 By the way, your thoughts about how golang handle the deps is almost correct 👍.

File: internal/lsp/elasticserver.go [hunk 2]
```diff
@@ -269,3 +295,175 @@ func getQName(ctx context.Context, f source.File, ident *source.IdentifierInfo,
...
+	loc = strings.TrimPrefix(loc, "file://")
+	if strings.HasPrefix(loc, depsPath) {
+		strs := []string{pkgLoc.RepoURI, "blob", pkgLoc.Version, path}
+		return filepath.Join(strs...)
+	}
+	return "file://" + loc
+}
+
+// normalizePath trims the workspace folder prefix to get the file path in project.
+func normalizePath(path *string, cfg *packages.Config, pkgLoc *protocol.PackageLocator, depsString string) {
```
Comment (Poytr1) [inline internal/lsp/elasticserver.go]: Can you add some unit tests for string normalization here? 😕

Comment (movie-travel-code) [reply internal/lsp/elasticserver.go]: Of course, I will give it a try!

File: internal/lsp/elasticserver.go [hunk 3]
```diff
@@ -269,3 +295,175 @@ func getQName(ctx context.Context, f source.File, ident *source.IdentifierInfo,
...
+
+func getPkgVersionSlow(s *ElasticServer, dir string, cfg *packages.Config, pkgLoc *protocol.PackageLocator) {
+	// Note: 'goListDriver' doesn't provide the module information of the package. So we have to call 'go list'
+	// manually.
+	args := []string{"list", "-e", "-json", "-deps=true"}
+	args = append(args, "--", dir)
+
+	stdout := new(bytes.Buffer)
+	stderr := new(bytes.Buffer)
+	cmd := exec.Command("go", args...)
```
Comment (zfy0701) [inline internal/lsp/elasticserver.go]: this is assume the user already go installed on the machine. I would say not implement getPkgVersionSlow for now (just return a clear error message), add a todo to invoke correct go internal function to extract that info

Comment (movie-travel-code) [reply internal/lsp/elasticserver.go]: I agree with that call `go list` should be deprecated. Given that we only use `getPkgVersionFast`, I tightened the restrictions,  two valid revisions vs two `@`.

File: internal/lsp/elasticserver.go (hunk 2 above)
Comment (movie-travel-code) [reply internal/lsp/elasticserver.go]: done.

**Pull request merged**